### ISSUE TITLE
Revert to the original CHB or LRB

### ISSUE THREAD

**Issue by shnarazk** [shnarazk/splr#10]
# Objective

Splr-0.2.1's main ordering heuristics wasn't CHB! Bogus! I need to investigate CHB again with more clever eyes.

## Reference

- Liang, J. H., Ganesh, V., Poupart, P., & Czarnecki, K.: Exponential Recency Weighted Average Branching Heuristic for SAT Solvers, 2016.

> VSIDS decays the activities of all variables whereas CHB again only decays the Q scores of branched, propagated, and asserted variables.

- Liang, J. H., Ganesh, V., Poupart, P. & Czarnecki, K.: Learning Rate Based Branching Heuristic for SAT Solvers, 2016.

## branches

- https://github.com/shnarazk/splr/tree/20200126-var-reward
- https://github.com/shnarazk/splr/tree/20200121-RSR
- https://github.com/shnarazk/splr/tree/20200116-extreme-RSR
- https://github.com/shnarazk/splr/tree/20200111-LRB
- https://github.com/shnarazk/splr/tree/20200105-CHB
- https://github.com/shnarazk/splr/tree/20191230-CHB
- https://github.com/shnarazk/splr/tree/20191229-CHB

## So far

- Even after checking and modifying most part of the program, it still gets bad performance on UNSAT problems, compared with EVISDS. How about extending restart steps by multiplying with FOC, because UNSAT problems have (small) UNSAT cores that we should *focus* on? This also might be evidence that deep search works well.
- We can't change misc settings on **adaptive_restart** and **deep_search**. The present is ever the best. 
- Reason-Side Rewarding can improve performance.
- Tiny change rates of var activity are very good for 3SAT problems since they tend to fix the order of var selections. This means the effects of restart is reduced and so it's likely to execute exhaustive search instead of random jumping.

## Comparison target

dev-0.2.2

```
# ~/.cargo/bin/splr (splr 0.2.2-dev.0) @ 2020-01-05T10:37:14
solver,       num,                 target,    time
"splr",         1,           "UF250(100)", 144.184
"splr",         2,          "UUF250(100)", 361.036
"splr",         3, "3/SAT/v360-c1530/002", 195.049
"splr",         4, "3/SAT/v360-c1530/030", 243.745
"splr",         5, "3/SAT/v360-c1530/033",   4.581
"splr",         6, "3/SAT/v360-c1530/035",  17.230
"splr",         7, "3/SAT/v360-c1530/039",   ABORT
"splr",         8, "3/SAT/v360-c1530/051",   ABORT
"splr",         9, "3/SAT/v360-c1530/060",  20.544
"splr",        10, "3/SAT/v360-c1530/073", 228.022
"splr",        11, "3/SAT/v360-c1530/087",   0.811
"splr",        12, "3/SAT/v360-c1530/093",  29.456
"splr",        13, "3/UNS/v360-c1530/001", 474.862
"splr",        14, "3/UNS/v360-c1530/015", 257.642
"splr",        15, "3/UNS/v360-c1530/028", 433.232
"splr",        16, "3/UNS/v360-c1530/029",   ABORT
"splr",        17, "3/UNS/v360-c1530/031",   ABORT
"splr",        18, "3/UNS/v360-c1530/053", 449.915
"splr",        19, "3/UNS/v360-c1530/061",   ABORT
"splr",        20, "3/UNS/v360-c1530/086", 218.609
"splr",        21, "3/UNS/v360-c1530/089", 336.133
"splr",        22, "3/UNS/v360-c1530/096", 225.897
"splr",        23,          "SR2015/itox",   2.545
"splr",        24,          "SR2015/m283",  72.317
"splr",        25,           "SR2015/38b",   6.358
"splr",        26,           "SR2015/44b", 422.609
```

```
# ~/.cargo/bin/splr (splr 0.2.2-dev.0) @ 2019-12-29T09:21:39
solver,       num,                 target,    time
"splr",         1,           "UF250(100)", 120.159
"splr",         2,          "UUF250(100)", 367.286
"splr",         3, "3/SAT/v360-c1530/002",  91.467
"splr",         4, "3/SAT/v360-c1530/030", 278.417
"splr",         5, "3/SAT/v360-c1530/033",  11.541
"splr",         6, "3/SAT/v360-c1530/035",  14.478
"splr",         7, "3/SAT/v360-c1530/039",   ABORT
"splr",         8, "3/SAT/v360-c1530/051",   ABORT
"splr",         9, "3/SAT/v360-c1530/060",   ABORT
"splr",        10, "3/SAT/v360-c1530/073", 107.075
"splr",        11, "3/SAT/v360-c1530/087",  99.664
"splr",        12, "3/SAT/v360-c1530/093", 138.323
"splr",        13, "3/UNS/v360-c1530/001",   ABORT
"splr",        14, "3/UNS/v360-c1530/015", 327.836
"splr",        15, "3/UNS/v360-c1530/028", 486.770
"splr",        16, "3/UNS/v360-c1530/029",   ABORT
"splr",        17, "3/UNS/v360-c1530/031",   ABORT
"splr",        18, "3/UNS/v360-c1530/053", 377.727
"splr",        19, "3/UNS/v360-c1530/061",   ABORT
"splr",        20, "3/UNS/v360-c1530/086", 295.135
"splr",        21, "3/UNS/v360-c1530/089", 368.148
"splr",        22, "3/UNS/v360-c1530/096", 202.818
"splr",        23,          "SR2015/itox",   1.637
"splr",        24,          "SR2015/m283",  59.509
"splr",        25,           "SR2015/38b",  34.444
"splr",        26,           "SR2015/44b", 118.803
```

**Comment by shnarazk** [comment]
![スクリーンショット 2019-12-30 8 39 24](https://user-images.githubusercontent.com/997855/71564185-eec23500-2adf-11ea-91aa-690e65273e82.png)

**Comment by shnarazk** [comment]
## LRB

```rust
after_conflict_analysis() {
  LEARNT_COUNT += 1;
  for v in conflict_side + learnt_clause_vars {
    participated[v] += 1;
  }
  for v in learnt_clause_vars.map(|v| reason(v)).join() - clause_vars {
    reasoned[v] += 1;
  }
  if 0.06 < alpha {
    alpha -= 10E-6;
  }
}

fn on_assign(v: Var) {
  assigned[v] = LEARNT_COUNT;
  participated[v] = 0;
  reasoned[v] = 0;
}

fn on_unassign(v: Var) {
  let interval = LEARNT_COUNT - assigned[v];
  if 0 < interval {
    let r = participated[v] / interval;
    let s = reasoned[v] / interval;
    Q[v] = (1 - alpha) * Q[v] + alpha * r;
  }
}
```

**Comment by shnarazk** [comment]
It works well and landed on dev-0.2.2. Done.
